File: apps/cli/cmd/org.go
package cmd

import (
	"fmt"
	"os"
	"strings"

	"github.com/manifoldco/promptui"
	"github.com/spf13/cobra"
	"github.com/spf13/viper"
	"yishan/apps/cli/internal/api"
	"yishan/apps/cli/internal/config"
	"yishan/apps/cli/internal/output"
)

var orgListCmd = &cobra.Command{
	Use:   "list",
	Short: "List organizations",
	Long:  `List all organizations the authenticated user is a member of.`,
	Example: `  yishan org list
  yishan org list --output json`,
	RunE: func(cmd *cobra.Command, _ []string) error {
		verbose, err := cmd.Flags().GetBool("verbose")
		if err != nil {
			return err
		}

		response, err := apiClient.ListOrganizations()
		if err != nil {
			return err
		}

		renderData, err := toOrgListRenderData(response, verbose)
		if err != nil {
			return err
		}

		return output.PrintRenderData(renderData)
	},
}

var orgCreateCmd = &cobra.Command{
	Use:   "create",
	Short: "Create organization",
	Long:  `Create a new Yishan organization. The authenticated user becomes the first admin member.`,
	Example: `  yishan org create --name "My Org"
  yishan org create --name "My Org" --member-user-id uid1 --member-user-id uid2`,
	RunE: func(cmd *cobra.Command, _ []string) error {
		name, err := cmd.Flags().GetString("name")
		if err != nil {
			return err
		}
		memberUserIDs, err := cmd.Flags().GetStringSlice("member-user-id")
		if err != nil {
			return err
		}

		response, err := apiClient.CreateOrganization(api.CreateOrganizationInput{
			Name:          name,
			MemberUserIDs: memberUserIDs,
		})
		if err != nil {
			return err
		}

		return output.PrintAny(response)
	},
}

var orgDeleteCmd = &cobra.Command{
	Use:     "delete",
	Short:   "Delete organization",
	Long:    `Permanently delete an organization and all its projects and workspaces. This action cannot be undone.`,
	Example: `  yishan org delete --org-id <org-id>`,
	RunE: func(cmd *cobra.Command, _ []string) error {
		orgID, err := resolveOrgID(cmd)
		if err != nil {
			return err
		}

		response, err := apiClient.DeleteOrganization(orgID)
		if err != nil {
			return err
		}

		return output.PrintAny(response)
	},
}

var orgMemberAddCmd = &cobra.Command{
	Use:   "add",
	Short: "Add organization member",
	Long:  `Add a user to the current organization. Role defaults to "member".`,
	Example: `  yishan org member add --user-id <uid>
  yishan org member add --user-id <uid> --role admin`,
	RunE: func(cmd *cobra.Command, _ []string) error {
		orgID, err := resolveOrgID(cmd)
		if err != nil {
			return err
		}
		userID, err := cmd.Flags().GetString("user-id")
		if err != nil {
			return err
		}
		role, err := cmd.Flags().GetString("role")
		if err != nil {
			return err
		}

		response, err := apiClient.AddOrganizationMember(orgID, userID, role)
		if err != nil {
			return err
		}

		return output.PrintAny(response)
	},
}

var orgMemberRemoveCmd = &cobra.Command{
	Use:     "remove",
	Short:   "Remove organization member",
	Long:    `Remove a user from the current organization.`,
	Example: `  yishan org member remove --user-id <uid>`,
	RunE: func(cmd *cobra.Command, _ []string) error {
		orgID, err := resolveOrgID(cmd)
		if err != nil {
			return err
		}
		userID, err := cmd.Flags().GetString("user-id")
		if err != nil {
			return err
		}

		response, err := apiClient.RemoveOrganizationMember(orgID, userID)
		if err != nil {
			return err
		}

		return output.PrintAny(response)
	},
}

var orgCmd = &cobra.Command{
	Use:   "org",
	Short: "Organization operations",
	Long:  `Create, list, delete, and manage membership for Yishan organizations.`,
}

var orgMemberCmd = &cobra.Command{
	Use:   "member",
	Short: "Organization member operations",
	Long:  `Add and remove members from a Yishan organization.`,
}

var orgDefaultCmd = &cobra.Command{
	Use:   "default",
	Short: "Get or set the default organization",
	Long: `Print the default organization used by CLI commands, or set it with --org-id.

The default org is used whenever --org-id is not passed explicitly. It is
stored in context.yaml and is independent of which org is selected in the app.

Examples:
  yishan org default                      # print current default
  yishan org default --org-id <org-id>    # set default interactively
  yishan org default --org-id <org-id>    # set default to a specific org`,
	RunE: func(cmd *cobra.Command, _ []string) error {
		orgID, err := cmd.Flags().GetString("org-id")
		if err != nil {
			return err
		}

		// Set mode: --org-id provided.
		if strings.TrimSpace(orgID) != "" {
			orgID = strings.TrimSpace(orgID)
			if err := config.UpdateContext(appConfig.ContextPath, func(cfg *viper.Viper) {
				cfg.Set(config.KeyDefaultOrgID, orgID)
			}); err != nil {
				return err
			}
			appConfig.DefaultOrgID = orgID
			if !output.IsJSONOutput() {
				fmt.Printf("Default organization: %s\n", orgID)
				return nil
			}
			return output.PrintAny(map[string]string{"orgId": orgID, "status": "default"})
		}

		// Get mode: print the current default.
		if appConfig.DefaultOrgID == "" {
			return fmt.Errorf("no default org set: run `yishan org default --org-id <org-id>`")
		}

		if !output.IsJSONOutput() {
			fmt.Println(appConfig.DefaultOrgID)
			return nil
		}
		return output.PrintAny(map[string]string{"orgId": appConfig.DefaultOrgID})
	},
}

func selectOrganizationInteractive(cmd *cobra.Command) (string, error) {
	if stat, err := os.Stdin.Stat(); err != nil || (stat.Mode()&os.ModeCharDevice) == 0 {
		return "", fmt.Errorf("org-id is required: use --org-id <org-id>")
	}

	response, err := apiClient.ListOrganizations()
	if err != nil {
		return "", err
	}
	if len(response.Organizations) == 0 {
		return "", fmt.Errorf("no organizations available for your account")
	}

	items := make([]string, 0, len(response.Organizations))
	for _, organization := range response.Organizations {
		items = append(items, fmt.Sprintf("%s (%s)", organization.Name, organization.ID))
	}

	prompt := promptui.Select{
		Label: "Select organization",
		Items: items,
		Size:  12,
	}

	index, _, err := prompt.Run()
	if err != nil {
		return "", err
	}

	return response.Organizations[index].ID, nil
}

var orgClearCmd = &cobra.Command{
	Use:   "clear",
	Short: "Clear the default organization",
	RunE: func(_ *cobra.Command, _ []string) error {
		if err := config.UpdateContext(appConfig.ContextPath, func(cfg *viper.Viper) {
			cfg.Set(config.KeyDefaultOrgID, "")
		}); err != nil {
			return err
		}

		appConfig.DefaultOrgID = ""
		return output.PrintAny(map[string]string{"status": "cleared"})
	},
}

func init() {
	rootCmd.AddCommand(orgCmd)

	orgCmd.AddCommand(orgListCmd)
	orgCmd.AddCommand(orgCreateCmd)
	orgCmd.AddCommand(orgDeleteCmd)
	orgCmd.AddCommand(orgDefaultCmd)
	orgCmd.AddCommand(orgClearCmd)
	orgCmd.AddCommand(orgMemberCmd)
	orgMemberCmd.AddCommand(orgMemberAddCmd)
	orgMemberCmd.AddCommand(orgMemberRemoveCmd)

	orgCreateCmd.Flags().String("name", "", "organization name")
	orgCreateCmd.Flags().StringSlice("member-user-id", []string{}, "additional member user id")
	cobra.CheckErr(orgCreateCmd.MarkFlagRequired("name"))
	orgListCmd.Flags().BoolP("verbose", "v", false, "show full response fields")

	addOrgIDFlag(orgDeleteCmd)

	orgDefaultCmd.Flags().String("org-id", "", "organization ID to set as default")

	addOrgIDFlag(orgMemberAddCmd)
	orgMemberAddCmd.Flags().String("user-id", "", "member user ID")
	orgMemberAddCmd.Flags().String("role", "member", "member role (member|admin)")
	cobra.CheckErr(orgMemberAddCmd.MarkFlagRequired("user-id"))

	addOrgIDFlag(orgMemberRemoveCmd)
	orgMemberRemoveCmd.Flags().String("user-id", "", "member user ID")
	cobra.CheckErr(orgMemberRemoveCmd.MarkFlagRequired("user-id"))
}

File: apps/cli-rs/src/cmd/org.rs
use crate::cmd::resolve_org_id;
use crate::output::print_any;
use crate::runtime::AppRuntime;
use clap::{Args, Subcommand};
use dialoguer::Select;
use serde_json::json;

#[derive(Subcommand)]
pub enum OrgCommands {
    /// List organizations
    List(OrgListArgs),
    /// Create organization
    Create(OrgCreateArgs),
    /// Delete organization
    Delete(OrgDeleteArgs),
    /// Set current organization
    Use(OrgUseArgs),
    /// Show current organization
    Current,
    /// Clear current organization
    Clear,
    /// Organization member operations
    #[command(subcommand)]
    Member(OrgMemberCommands),
}

#[derive(Subcommand)]
pub enum OrgMemberCommands {
    /// Add a member
    Add(OrgMemberAddArgs),
    /// Remove a member
    Remove(OrgMemberRemoveArgs),
}

#[derive(Args)]
pub struct OrgListArgs {
    /// Show all fields
    #[arg(long, short = 'v')]
    pub verbose: bool,
}

#[derive(Args)]
pub struct OrgCreateArgs {
    #[arg(long)]
    pub name: String,
    #[arg(long = "member-user-id")]
    pub member_user_ids: Vec<String>,
}

#[derive(Args)]
pub struct OrgDeleteArgs {
    #[arg(long)]
    pub org_id: Option<String>,
}

#[derive(Args)]
pub struct OrgUseArgs {
    /// Organization ID to activate
    #[arg(long)]
    pub org_id: Option<String>,
}

#[derive(Args)]
pub struct OrgMemberAddArgs {
    #[arg(long)]
    pub org_id: Option<String>,
    #[arg(long)]
    pub user_id: String,
    #[arg(long, default_value = "member")]
    pub role: String,
}

#[derive(Args)]
pub struct OrgMemberRemoveArgs {
    #[arg(long)]
    pub org_id: Option<String>,
    #[arg(long)]
    pub user_id: String,
}

pub async fn run(cmd: OrgCommands, runtime: &AppRuntime) -> anyhow::Result<()> {
    let client = runtime.api_client();

    match cmd {
        OrgCommands::List(args) => {
            let resp = client.list_organizations().await?;
            if args.verbose {
                print_any(resp)
            } else {
                let rows: Vec<serde_json::Value> = resp
                    .organizations
                    .iter()
                    .map(|o| {
                        json!({
                            "id": o.id,
                            "name": o.name,
                            "members": o.members.len(),
                            "createdAt": o.created_at,
                        })
                    })
                    .collect();
                print_any(rows)
            }
        }

        OrgCommands::Create(args) => {
            let resp = client
                .create_organization(&args.name, args.member_user_ids)
                .await?;
            print_any(resp)
        }

        OrgCommands::Delete(args) => {
            let org_id = resolve_org_id(args.org_id.as_deref(), runtime)?;
            let resp = client.delete_organization(&org_id).await?;
            print_any(resp)
        }

        OrgCommands::Use(args) => {
            let org_id = if let Some(id) = args.org_id.filter(|s| !s.is_empty()) {
                id
            } else {
                select_org_interactive(&client).await?
            };
            runtime.set_current_org(&org_id)?;
            print_any(json!({ "orgId": org_id, "status": "active" }))
        }

        OrgCommands::Current => {
            let cfg = runtime.config();
            if cfg.current_org_id.is_empty() {
                anyhow::bail!("no active org: run `yishan org use --org-id <org-id>`");
            }
            let resp = client.list_organizations().await?;
            let org = resp
                .organizations
                .into_iter()
                .find(|o| o.id == cfg.current_org_id)
                .ok_or_else(|| {
                    anyhow::anyhow!(
                        "current org {} not found in accessible organizations",
                        cfg.current_org_id
                    )
                })?;
            print_any(json!({
                "id": org.id,
                "name": org.name,
                "createdAt": org.created_at,
                "members": org.members,
            }))
        }

        OrgCommands::Clear => {
            runtime.set_current_org("")?;
            print_any(json!({ "status": "cleared" }))
        }

        OrgCommands::Member(sub) => match sub {
            OrgMemberCommands::Add(args) => {
                let org_id = resolve_org_id(args.org_id.as_deref(), runtime)?;
                let resp = client
                    .add_organization_member(&org_id, &args.user_id, &args.role)
                    .await?;
                print_any(resp)
            }
            OrgMemberCommands::Remove(args) => {
                let org_id = resolve_org_id(args.org_id.as_deref(), runtime)?;
                let resp = client
                    .remove_organization_member(&org_id, &args.user_id)
                    .await?;
                print_any(resp)
            }
        },
    }
}

async fn select_org_interactive(client: &crate::api::ApiClient) -> anyhow::Result<String> {
    if !atty::is(atty::Stream::Stdin) {
        anyhow::bail!("org-id is required: use --org-id <org-id>");
    }
    let resp = client.list_organizations().await?;
    if resp.organizations.is_empty() {
        anyhow::bail!("no organizations available for your account");
    }
    let items: Vec<String> = resp
        .organizations
        .iter()
        .map(|o| format!("{} ({})", o.name, o.id))
        .collect();
    let idx = Select::new()
        .with_prompt("Select organization")
        .items(&items)
        .interact()?;
    Ok(resp.organizations[idx].id.clone())
}
